[FIRRTL] Parse CircuitTarget and ModuleTarget Annotations

seldridge opened the pull request:
Adds the ability to parse and store CircuitTarget or ModuleTarget
local Annotations as MLIR Attributes.  Annotations are stringly-typed
in an array of attributes.  The "target" field is removed
as part of this process as this is now implicit based on the operation
to which the Attribute is applied.
    
Any other types of Annotations (e.g., ComponentTarget or non-local)
produce a runtime error with JSON diagnostic information printing.
    
The utility firtool is updated to take a --annotation-file argument
which can be used to import annotations from an annotation file.
Additionally, a non-standard inline annotation syntax is added.

Annotations ingested via --annotation-file are appended to existing inline annotations. 

### Example 1: Input Annotation File

Given the following circuit:

```scala
circuit Foo:
  module Bar:
    input a: UInt<1>
    output b: UInt<1>

    b <= a
  module Foo:
    input a: UInt<1>
    output b: UInt<1>

    inst bar of Bar
    bar.a <= a
    b <= bar.b
```

And the following annotation file:
```json
[
  {
    "class":"firrtl.passes.InlineAnnotation",
    "target":"Foo.Bar"
  },
  {
    "class":"firrtl.transforms.NoDedupAnnotation",
    "target":"~Foo|Bar"
  },
  {
    "class":"firrtl.stage.RunFirrtlTransformAnnotation",
    "transform":"firrtl.passes.InlineInstances"
  }
]
```

You can import this into CIRCT using:

```bash
firtool Foo.fir --annotation-file Foo.anno.json
```

To get a bunch of `#firrtl.*` attributes:

```mlir
module  {
  firrtl.circuit "Foo" attributes {annotations = [{class = "firrtl.stage.RunFirrtlTransformAnnotation", transform = "firrtl.passes.InlineInstances"}]} {
    firrtl.module @Bar(%a: !firrtl.uint<1>, %b: !firrtl.flip<uint<1>>) attributes {annotations = [{class = "firrtl.passes.InlineAnnotation"}, {class = "firrtl.transforms.NoDedupAnnotation"}]} {
      firrtl.connect %b, %a : !firrtl.flip<uint<1>>, !firrtl.uint<1>
    }
    firrtl.module @Foo(%a: !firrtl.uint<1>, %b: !firrtl.flip<uint<1>>) {
      %bar_a, %bar_b = firrtl.instance @Bar {name = "bar", portNames = ["a", "b"]} : !firrtl.flip<uint<1>>, !firrtl.uint<1>
      firrtl.connect %bar_a, %a : !firrtl.flip<uint<1>>, !firrtl.uint<1>
      firrtl.connect %b, %bar_b : !firrtl.flip<uint<1>>, !firrtl.uint<1>
    }
  }
}
```

The above can be round tripped into/out of an `*.mlir` file.

### Example 2: Inline Annotation Syntax

The above can be represented using a non-standard _inline annotation syntax_. Here, some JSON is packed into a `%[...]` area after the circuit. I'm working with the other Chisel/FIRRTL developers to see about getting something like this supported by the Scala FIRRTL Compiler and into the spec. I expect that this will change in the future and we can add support for the new syntax once consensus is reached.

Example:

```scala
circuit Bar: %[[{"class":"Foo"},{"class":"Bar","target":"~Bar|Bar"}]] @[Bar.scala 42:42]
  module Bar: @[Bar.scala 43:43]
    skip
```

This then parses into:

```mlir
module  {
  firrtl.circuit "Bar" attributes {annotations = [{class = "Foo"}]} {
    firrtl.module @Bar() attributes {annotations = [{class = "Bar"}]} {
    }
  }
}
```

### Example 3: Invalid JSON

If there is actual invalid JSON, this should be reported correctly. Due to the fact that `llvm::json::ParseError` does not have public line/column/offset members, the error can't be great. (If this was public, then we should be able to integrate this with a good source manager location error.)

Consider the following which includes invalid JSON (I used a `,` instead of a `:`):

```scala
circuit Foo: %[[{"a","a"}]]
  module Foo:
    skip
```

This will error like so:

```
annotations/ParserFailure.fir:1:14: error: Failed to parse inline JSON annotations
circuit Foo: %[[{"a","a"}]]
             ^
annotations/ParserFailure.fir:1:14: note: [1:6, byte=6]: Expected : after object key
```

### Example 4: Invalid/Unsupported Annotation JSON

This PR only adds support for annotations without a target (some subclass of [`firrtl.annotations.NoTargetAnnotation`](https://www.chisel-lang.org/api/firrtl/1.4.2/firrtl/annotations/NoTargetAnnotation.html)) or annotations targeting the entire circuit or a specific module. This PR treats anything else as an error.

This strict requirement can be lessened by, in the future, increasing support or by adding an option that promotes other annotations to target the entire circuit or drops them. I wanted to start with the strict option and expand functionality from there.

Additionally, annotations need to be an array of objects. Anything else is invalid.

Consider the following which includes a reference target, `~Baz|Baz>a`:

```scala
circuit Baz: %[[{"hello":"world","target":"~Baz|Baz>a"}]]
  module Baz:
    input a: UInt<1>
    output b: UInt<1>

    b <= a
```

When compiled this produces the following error:

```
annotations/ReferenceTarget.fir:1:14: error: Invalid/unsupported annotation format
circuit Baz: %[[{"hello":"world","target":"~Baz|Baz>a"}]]
             ^
annotations/ReferenceTarget.fir:1:14: note: See inline comments for problem area in JSON:
[
  {
    "hello": "world",
    "target": /* error: Unsupported target (not a CircuitTarget or ModuleTarget) */ "~Baz|Baz>a"
  }
]
```

### TODO

- [x] Figure out how to represent annotations inline in a `*.fir` file
- [x] Add tests
- [x] Use better error message reporting (emit an error on the circuit op and/or investigate the ability of `llvm::json::Path` to point to a specific problematic Annotation in the JSON
- [x] Switch to stringly-type `DictionaryAttr` to store arbitrary annotations

lattner commented on include/circt/Dialect/FIRRTL/FIRRTLStructure.td:
nit: 80 columns :-)

seldridge replied:
Change in a fixup commit.

lattner commented on lib/Dialect/FIRRTL/Import/FIRParser.cpp:
```diff
@@ -59,6 +62,98 @@ static StringAttr filterUselessName(StringAttr name) {
+          return failure();
+          break;
+        case UnknownAnnotationID:
+          llvm::errs() << "Unknown annotation: " << className << "\n";
```
The JSON parser doesn't support tracking file locations at all ;-( but it is untidy to emit errors with `errs`.  I'd recommend passing a location of the circuit declaration in the .fir file or something, and then emitError'ing on it.  This will allow us to filecheck these things etc.

seldridge replied:
I can switch to that. I'll probably just parse the circuit definition, parse the annotation file (erroring pointing at the circuit location) to build up the annotation data structure, then proceed to parse everything.

seldridge replied:
Fixed in newer squash commits. The PR now delays processing the input annotation file until it has a circuit. The error is now associated with the circuit op. A JSON parser error failure isn't amazing:

lattner commented on lib/Dialect/FIRRTL/Import/FIRParser.cpp:
```diff
@@ -59,6 +62,98 @@ static StringAttr filterUselessName(StringAttr name) {
+        return failure();
+      }
+    } else {
+      return failure();
```
Should these failure cases emit an error?

I'd also recommend inverting the logic to reduce nesting here, e.g.:

```
auto object = value.getAsObject();
if (!object) {
  stuff
  return;
}
```

seldridge replied:
Cleaned up in squash commits.

lattner commented on lib/Dialect/FIRRTL/Import/FIRParser.cpp:
```diff
@@ -2711,19 +2830,31 @@ ParseResult FIRCircuitParser::parseCircuit() {
+  if (annotations) {
+    for (auto a : *annotations.getValue().getAsArray()) {
+      if (failed(fromJSON(a, annotationMap, root, context))) {
+        llvm::errs() << "Failed to parse annotation file...\n";
```
Please use emitError on a location in the fir file.

seldridge replied:
Fixed in squash commits.

lattner commented on tools/firtool/firtool.cpp:
```diff
@@ -112,7 +118,7 @@ processBuffer(std::unique_ptr<llvm::MemoryBuffer> ownedBuffer,
     firrtl::FIRParserOptions options;
     options.ignoreInfoLocators = ignoreFIRLocations;
-    module = importFIRRTL(sourceMgr, &context, options);
+    module = importFIRRTL(sourceMgr, annotations, &context, options);
```
The top level dance here is a bit weird to me because both the caller and callee of importFIRRTL has to know about JSON stuff.  I'd recommend that firtool.cpp pass in annotations as a filename, or StringRef for the buffer, and put all the JSON parsing stuff (JSON.h etc) into the caller.

Per the macro comment, if might make sense to make annotation parsing be a separate call from importFIRRTL.

seldridge replied:
I think this is all cleaned up. Now `firtool` no longer has to know about JSON. The actual annotation file is pulled in by the source manager in the parser.

lattner commented:
This is looking really great.  Per our discussion yesterday, I'd recommend a couple macro things:

1) Start with "stringly typed" annotations instead of defined-in-tblgen attributes, so we can scoop up the generality of all annotations at once.  Having a first class `Annotation` type in C++ makes sense, it can just contain strings instead of knowing about specific kinds of annotation like `InlineAnnotation`.  OTOH, you could go all the way untyped and just use an MLIR arrayattr of key/value pairs or something too.

2) This logic is all quite different than the mechanics of parsing a .fir file.  I'd split it out to its own .cpp file, and potentially even consider making this be a utility function that is called separate from the FIRParser.

3) Please think about how to write tests for this, e.g. how to test incorrect annotations etc.  Having an `--annotation-file` command to firtool is awesome, but it isn't going to to work well with tests that are -split-input-file.  Maybe we can invent a special syntax or directive in the fir file?

seldridge commented:
This is not ready for a review, yet, but I wanted to provide the update that this has moved away from custom FIRRTL attributes to store annotations and is now just using an `ArrayAttr` of `DictionaryAttr` due to the need to handle arbitrary, unknown annotations. This representation (and the fact that Attributes and JSON are essentially interchangeable) means that this PR should be able to ingest all Annotations at one fell swoop.

seldridge commented:
After talking with @jackkoenig, he suggested an inline annotation syntax like the following:

```scala
circuit Bar: %[[{"class":"Foo"},{"class":"Bar","target":"~Bar|Bar"}]]
  module Bar:
    skip
```

This is basically the same style as how source locators are encoded using `@[...]`, but it sticks a JSON snippet in the brackets and uses a different leading character: `%[...]`. Alternatives considered:

JSON at the top:
```scala
[{"class":"Foo"},{"class":"Bar","target":"~Bar|Bar"}]
circuit Bar: 
  module Bar:
    skip
```

JSON in a comment at the top:
```scala
; [{"class":"Foo"},{"class":"Bar","target":"~Bar|Bar"}]
circuit Bar: 
  module Bar:
    skip
```

Any of the above would all parse into:

```mlir
module  {
  firrtl.circuit "Bar" attributes {annotations = [{class = "Foo"}]} {
    firrtl.module @Bar() attributes {annotations = [{class = "Bar"}]} {
    }
  }
}
```

I'm intending for this to only be legal for the circuit statement. Any place else, e.g., on a module, would be an error. This then defines two ways of adding annotations:

1. `--annotation-file`
2. Inline with the circuit `%[...]`

These would compose via appending, e.g., using (1) and (2) would result in the annotations from both showing up. (This is not intended to be a mode people use, but some behavior should be defined.) Alternatively, using both could be an error.

lattner commented on include/circt/Dialect/FIRRTL/FIRAnnotations.h:
```diff
@@ -0,0 +1,31 @@
+namespace firrtl {
+
+bool fromJSON(llvm::json::Value &value,
+              llvm::StringMap<llvm::SmallVector<DictionaryAttr>> &annotationMap,
```
Per offline discussion, plz change this to 
`llvm::StringMap<ArrayAttr> &annotationMap,`

seldridge replied:
Changed in a later `!squash` commit.

lattner commented on include/circt/Dialect/FIRRTL/FIRRTLStructure.td:
This can take an ArrayAttr instead of ArrayRef<dict>

seldridge replied:
Thanks. This is cleaner. I applied this suggestion in a later `!squash` commit.

lattner commented on lib/Dialect/FIRRTL/FIRRTLDialect.cpp:
```diff
@@ -13,6 +13,7 @@
 #include "circt/Dialect/FIRRTL/FIRRTLDialect.h"
 #include "circt/Dialect/FIRRTL/FIRRTLOps.h"
 #include "mlir/IR/DialectImplementation.h"
+#include "llvm/ADT/TypeSwitch.h"
```
I don't think this #include is used.

seldridge replied:
Fixed in later commits.

lattner commented on lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:
```diff
@@ -0,0 +1,181 @@
+
+/// Deserialize a JSON value into a Target-keyed array of Annotations
+/// (represented as DictionaryAttrs).
+bool fromJSON(llvm::json::Value &value,
```
You can remove the namespace scoping on line 21/22 and declare this as:

bool circt::firrtl::fromJSON(llvm::json::Value &value, ...

seldridge replied:
Flattened this out as suggested.

lattner commented on lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:
```diff
@@ -0,0 +1,181 @@
+    } else {
+      // This is a legacy target using the firrtl.annotations.Named type.  This
+      // can be trivially canonicalized to a non-legacy target, so we do it with
+      // the following thre mappings:
```
typo thre -> three

seldridge replied:
Good catch. Fixed.

lattner commented on lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:
```diff
@@ -0,0 +1,181 @@
+    // Null
+    if (auto a = value.getAsNull()) {
+      p.report("Found a null in the JSON, returning 'nullptr'");
+      return nullptr;
```
You could represent this as "unit" the empty tuple attr if you wanted to.  Ths would make it so anything would round trip and would fix the missing null error handling at the callers of JSONToAttribute

seldridge replied:
Changed to `{}`.

lattner commented on lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:
```diff
@@ -0,0 +1,181 @@
+
+  /// Convert arbitrary JSON to an MLIR Attribute.
+  std::function<Attribute(llvm::json::Value &, llvm::json::Path)>
+      JSONToAttribute =
```
nit plz name functions/variables with lower case letter, maybe `convertJSONToAttribute`

seldridge replied:
Changed to use the `convertJSONToAttribute` suggestion.

seldridge commented:
Changes since last review:

1. Annotation representation is changed from `SmallVector<DictionaryAttr>` to `ArrayAttr`. 
2. Tests are added for `--annotation-file` with `firtool`.
3. `--annotation-file` annotations are _appended_ after inline annotations (if they exist).

Some of the legwork related to (1) is a little annoying when you want to mutate the `ArrayAttr`. I did this by building a new, mutable array, then creating a new `ArrayAttr`. There may be a better way to do this.

lattner commented on include/circt/Dialect/FIRRTL/FIRAnnotations.h:
```diff
@@ -0,0 +1,31 @@
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/Support/JSON.h"
```
Please forward declare things in headers to reduce transitive includes, e.g.:

`namespace llvm { namespace json { class Path; class Value; }}`
`namespace mlir { class ArrayAttr; }`

It looks like StringMap isn't in LLVM.h, so StringMap.h is ok.

seldridge replied:
Fixed in rebase.

lattner commented on include/circt/Dialect/FIRRTL/FIRAnnotations.h:
```diff
@@ -0,0 +1,31 @@
+//===- FIRAnnotations.h - .fir to FIRRTL dialect parser -------------------===//
```
Plz change the header comment line to be more specific.

seldridge replied:
Fixed in the rebase.

lattner commented on include/circt/Dialect/FIRRTL/FIRRTLStructure.td:
Instead of making the annotations array optional, did you consider representing this state with an empty array?  That reduces the encoding ambiguity between `Some([])` and `None`, and will reduce some annoyance dealing with the optional.  It would be important not to print the empty attribute array in the asmsyntax tho

seldridge replied:
Good idea. This makes more sense with the representation you suggest. I tried this and it proved weird where tblgen wasn't cooperative. I'm going to look at this tomorrow and see if I can get it to work. If it looks like an upstream issue, I'll see about a fix, but may opt to merge this as is and correct later.

seldridge replied:
I'll look at an upstream fix for this. `ArrayAttr` doesn't implement the `constBuilderCall` method which is needed to generate a `DefaultValuedAttr`. There's a kludge to get around this added in 06b2fe4b6aaacc6c74d9c3c8700bdff283765192.

lattner commented on lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:
```diff
@@ -0,0 +1,182 @@
+//===- FIRAnnotations.cpp - .fir to FIRRTL dialect parser -----------------===//
```
Plz update comment line, this is about annotations not the parser as a whole

seldridge replied:
Fixed in rebase.

lattner commented on lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:
```diff
@@ -0,0 +1,182 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// Provide utilities related to dealing with Scala FIRRTL Compiler Annotations.
```
I'd drop "scala" as irrelevant

seldridge replied:
Fixed in rebase.

lattner commented on lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:
```diff
@@ -0,0 +1,182 @@
+#include "circt/Dialect/FIRRTL/FIRAnnotations.h"
+
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/OperationSupport.h"
```
I'd consider a "using json = llvm::json;" to simplify the qualifications below. 

"using namespace" things in headers are evil, but they are ok in .cpp files.

seldridge replied:
Good idea. Fixed in the rebase.

lattner commented on lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:
```diff
@@ -0,0 +1,182 @@
+      NamedAttrList metadata;
+      for (auto b : *a)
+        metadata.append(b.first,
+                        convertJSONToAttribute(b.second, p.field(b.first)));
```
This isn't handling the null return of `convertJSONToAttribute`.  I'd recommend eliminating it as a possibility by representing JSON `Null` as `UnitAttr::get(context)` since tuples aren't otherwise used.

seldridge replied:
Fixed in the rebase and added a test that all possible JSON values can be read in.

lattner commented on lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:
```diff
@@ -0,0 +1,182 @@
+  };
+
+  // The JSON value must be an array of objects.  Anything else is reported as
+  // inavlid.
```
typo inavlid

seldridge replied:
Fixed in later `!fixup` commit.

lattner commented on lib/Dialect/FIRRTL/Import/FIRLexer.cpp:
```diff
@@ -279,6 +284,38 @@ FIRToken FIRLexer::lexFileInfo(const char *tokStart) {
+      if (curPtr - 1 != curBuffer.end())
+        break;
+      LLVM_FALLTHROUGH;
+    case '\n': // Vertical whitespace isn't allowed in inline annotations.
```
Do you want to eventually support escapes in these annotations?  If so, it is probably best to reject `\\` as an error for now.

seldridge replied:
I beefed this up in a `!squash` commit. I'm handling escapes and statefully parsing `"string"` expressions now. This likely has bugs, but is better than it was.

lattner commented on lib/Dialect/FIRRTL/Import/FIRParser.cpp:
```diff
@@ -25,6 +26,10 @@
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringSwitch.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/JSON.h"
```
Please prune unneeded #includes.  StringMap comes in from FIRAnnotations, StringSwitch.h isn't used, Error.h may already come in, etc.

seldridge replied:
Cleaned up in a `!fixup` commit. There are a few more extraneous ones in this file that I didn't add that I can rip out on the `main` branch.

lattner commented on lib/Dialect/FIRRTL/Import/FIRParser.cpp:
```diff
@@ -740,6 +797,88 @@ ParseResult FIRParser::parseOptionalRUW(RUWAttr &result) {
   return success();
 }
 
+ParseResult FIRParser::importAnnotations(const SMLoc &loc,
```
SMLoc is trivial and can be passed by value instead of `const&`.

seldridge replied:
Good point. Fixed in a `!fixup` commit.

lattner commented on lib/Dialect/FIRRTL/Import/FIRParser.cpp:
```diff
@@ -740,6 +797,88 @@ ParseResult FIRParser::parseOptionalRUW(RUWAttr &result) {
 
+ParseResult FIRParser::importAnnotations(const SMLoc &loc,
+                                         StringRef annotationsStr,
+                                         MLIRContext *context,
```
No need to pass in a context, you can call `getContext()` in FIRParser.

seldridge replied:
Nice catch. Updated in a `!fixup` commit.

lattner commented on lib/Dialect/FIRRTL/Import/FIRParser.cpp:
```diff
@@ -740,6 +797,88 @@ ParseResult FIRParser::parseOptionalRUW(RUWAttr &result) {
+ParseResult FIRParser::importAnnotations(const SMLoc &loc,
+                                         StringRef annotationsStr,
+                                         MLIRContext *context,
+                                         StringRef errorMsgParse) {
```
I don't think you need to pass in errorMsgParse since you have location info that makes it obvious what sort of thing you're parsing.  I'd forward substitute it to simplify the code a bit.

seldridge replied:
I'll opt for a more a generic message here that covers both the file and inline annotation situation. This can be improved, e.g., with an actual SMLoc pointing at the annotation file at some later point. Changed in a `!fixup` commit.

lattner commented on lib/Dialect/FIRRTL/Import/FIRParser.cpp:
```diff
@@ -740,6 +797,88 @@ ParseResult FIRParser::parseOptionalRUW(RUWAttr &result) {
+
+  llvm::json::Path::Root root;
+  llvm::StringMap<ArrayAttr> annotationMap;
+  if (annotations) {
```
I'd hoist the "if annotations" check up a bit, invert it and make it an early out to simplify/unnest the logic:

`if (!annotations) return success();`

seldridge replied:
Good thought. Looking at this again, this can actually just be removed. This is already being checked for errors above and blind `annotations.get()` will succeed, unless I misunderstand how `llvm::Expected` works...

lattner commented on lib/Dialect/FIRRTL/Import/FIRParser.cpp:
```diff
@@ -740,6 +797,88 @@ ParseResult FIRParser::parseOptionalRUW(RUWAttr &result) {
+
+  for (auto a : annotationMap.keys()) {
+    if (!state.annotationMap[a]) {
+      state.annotationMap[a] = annotationMap[a];
```
This is doing a double lookup, once to check the entry and once to insert it.  The insert method has a form that returns an iterator and a bool "insert succeeded" but in this case I'd recommend:

```
auto &entry = state.annotationMap[a];
if (!entry) {
  entry = annotationMap[a];
  continue;
}
```

... work with `entry` and `annotationMap[a]`

seldridge replied:
Nice! Switching to this.

lattner commented on lib/Dialect/FIRRTL/Import/FIRParser.cpp:
```diff
@@ -740,6 +797,88 @@ ParseResult FIRParser::parseOptionalRUW(RUWAttr &result) {
+    for (auto a : state.annotationMap[a])
+      annotationVec.push_back(a);
+    for (auto a : annotationMap[a])
+      annotationVec.push_back(a);
```
Depending on how the overloads work, you might be able to use:

`annotationVec.append(annotationMap[a]);`

etc, but maybe not.

seldridge replied:
The only variant that seemed to work here is the `iter_start, iter_end`. It doesn't save any lines, but it's clearer than loops. Good suggestion.

lattner commented on lib/Dialect/FIRRTL/Import/FIRParser.cpp:
```diff
@@ -740,6 +797,88 @@ ParseResult FIRParser::parseOptionalRUW(RUWAttr &result) {
+                                            MLIRContext *context) {
+
+  ParseResult result = success();
+  if (state.annotationsBuf)
```
I'd recomment inverting + early exit to simplify the logic:

`if (!state.annotationsBuf) return success();`

seldridge replied:
Fixed in a later commit.

seldridge replied:
Fixed in a later commit.

lattner commented on lib/Dialect/FIRRTL/Import/FIRParser.cpp:
```diff
@@ -740,6 +797,88 @@ ParseResult FIRParser::parseOptionalRUW(RUWAttr &result) {
+  }
+
+  // Input annotations is non-empty.  Exit quickly if the target doesn't exist.
+  // Otherwise, construct anew ArrayAttr that includes existing and new
```
typo anew

seldridge replied:
Fixed in later `!fixup` commit.

lattner reviewed (approved):
This patch is really great, thank you for shifting the array processing stuff.  There are a bunch of nit-picky things below, but feel free to land this after working through them.  Great work @seldridge !
